### PR TITLE
ENH: interpolate.FloaterHormaninterpolator: add array API standard support

### PR DESCRIPTION
**Pull request by prady0t**
Adding array api support for BarycentricInterpolator and FloaterHormannInterpolator

#### Reference issue
Related to: #23754 

#### AI Generation Disclosure

Some code blocks were generated by ChatGPT which I modified myself (since they were not so useful), some np to xp conversion methods were suggested by AI, some I found online.

### CODE REVIEW

**Comment by prady0t** [comment]
See: https://github.com/scipy/scipy/issues/23754#issuecomment-4366851936 for context.

File: scipy/interpolate/_bary_rational.py [hunk 1]
```diff
@@ -29,37 +29,43 @@
...
-        z = np.asarray(x)
-        f = np.asarray(y)
+        # Convert inputs
+        z = xp.asarray(x)
+        f = xp.asarray(y)
 
         self._input_validation(z, f, **kwargs)
 
-        f = np.moveaxis(f, self._axis, 0)
+        f = xp.moveaxis(f, axis, 0)
```
Comment (tylerjereddy) [inline scipy/interpolate/_bary_rational.py]: What does the `self._axis` -> `axis` mutation have to do with array API support?

File: scipy/interpolate/_bary_rational.py [hunk 2]
```diff
@@ -29,37 +29,43 @@
 
 import numpy as np
 import scipy
-
+from scipy._lib._array_api import array_namespace, xp_size, xp_capabilities
 
 __all__ = ["AAA", "FloaterHormannInterpolator"]
 
-
+@xp_capabilities()
```
Comment (tylerjereddy) [inline scipy/interpolate/_bary_rational.py]: interesting, should the decorator go on the "_private" base class?

File: scipy/interpolate/tests/test_bary_rational.py [hunk 3]
```diff
@@ -315,72 +319,81 @@ def test_iv(self):
...
+    def test_complex(self, xp):
+        x = xp.linspace(-1, 1, num=50)
         z = x + x*1j
-        r = FloaterHormannInterpolator(z, np.sin(z), d=12)
-        xx = np.linspace(-1, 1, num=1000)
+        r = FloaterHormannInterpolator(z, xp.sin(z), d=12)
+        xx = xp.linspace(-1, 1, num=1000)
         zz = xx + xx*1j
-        assert_allclose(r(zz), np.sin(zz), rtol=1e-12)
+        assert_allclose(r(zz), xp.sin(zz), rtol=1e-12)
```
Comment (tylerjereddy) [inline scipy/interpolate/tests/test_bary_rational.py]: Here and elsewhere--I think there may be some inconsistency in only sometimes converting over to the `xp_assert` functions, which you probably should do when swapping over to the array API style tests.

File: scipy/interpolate/_bary_rational.py [hunk 4]
```diff
@@ -97,40 +103,48 @@ def __call__(self, z):
...
+            CC = xp.divide(1, diff)
             # Vector of values
             r = CC @ (weights * support_values) / (CC @ weights)
 
         # Deal with input inf: `r(inf) = lim r(z) = sum(w*f) / sum(w)`
-        if np.any(np.isinf(zv)):
-            r[np.isinf(zv)] = (np.sum(weights * support_values)
-                               / np.sum(weights))
+        if xp.any(xp.isinf(zv)):
+            r[xp.isinf(zv)] = (xp.sum(weights * support_values)
```
Comment (tylerjereddy) [inline scipy/interpolate/_bary_rational.py:133]: Here and elsewhere--I think JAX arrays are immutable so either you'll want to deal with lazy arrays or perhaps start off by disabling JIT support in `xp_capabilities`

**Comment by tylerjereddy** [comment]
You'll probably want to fill in the AI usage disclosure in the OC as well--that's particularly important for first time contributors, but we're all trying to do it now for transparency.

File: scipy/interpolate/_bary_rational.py [hunk 5]
```diff
@@ -97,40 +103,48 @@ def __call__(self, z):
             Input values.
         """
         # evaluate rational function in barycentric form.
-        z = np.asarray(z)
-        zv = np.ravel(z)
+        xp = array_namespace(z)
+        z = xp.asarray(z)
+        zv = xp.reshape(z, (-1,))
```
Comment (lucascolley) [inline scipy/interpolate/_bary_rational.py]: we do have an `xp_ravel` in `_lib._array_api` which might make the diff slightly more clear

**Comment by j-bowhay** [comment]
I am quite concerned am the potential use of AI here since the title and pr description is wrong, this pr doesn't touch `BarycentricInterpolator`

**Comment by prady0t** [comment]
> I am quite concerned am the potential use of AI here since the title and pr description is wrong, this pr doesn't touch BarycentricInterpolator

It was meant for the future work in this PR itself 😅

File: scipy/interpolate/_bary_rational.py (hunk 1 above)
Comment (prady0t) [reply scipy/interpolate/_bary_rational.py]: Must have done it accidentally.

File: scipy/interpolate/_bary_rational.py (hunk 2 above)
Comment (prady0t) [reply scipy/interpolate/_bary_rational.py]: Actually, it's not. Thanks for pointing out.

File: scipy/interpolate/tests/test_bary_rational.py (hunk 3 above)
Comment (prady0t) [reply scipy/interpolate/tests/test_bary_rational.py]: This function doesn't complain, so yes, I can safely test with `xp_assert`. Some test functions need a lot more work, though. For example, the `test_polyinterp` relies on `BarycentricInterpolator` but it's not xp compatible yet; `test_zeros`, ' test_no_poles `, test_batch` depend on scipy.linalg, which needs to be dealt with first.

File: scipy/interpolate/_bary_rational.py (hunk 5 above)
Comment (prady0t) [reply scipy/interpolate/_bary_rational.py]: Thanks for pointing out!

**Comment by j-bowhay** [comment]
> > I am quite concerned am the potential use of AI here since the title and pr description is wrong, this pr doesn't touch BarycentricInterpolator
> 
> It was meant for the future work in this PR itself 😅

I would stick to doing one at a time

**Comment by prady0t** [comment]
> I would stick to doing one at a time

Should I open another PR for Barycentricinterpolator?

**Comment by j-bowhay** [comment]
> > I would stick to doing one at a time
> 
> Should I open another PR for Barycentricinterpolator?

I'd wait until this is finished and merged. Also, the code you are working on here is completely unrelated to `BarycentricInterpolator`

File: scipy/interpolate/_bary_rational.py [hunk 6]
```diff
@@ -29,37 +29,41 @@
...
 
         self._input_validation(z, f, **kwargs)
 
-        f = np.moveaxis(f, self._axis, 0)
+        f = xp.moveaxis(f, self._axis, 0)
 
         # Remove infinite or NaN function values and repeated entries
-        to_keep = np.logical_and.reduce(
-            ((np.isfinite(f)) & (~np.isnan(f))).reshape(f.shape[0], -1),
+        reshaped = xp.reshape(xp.isfinite(f), (f.shape[0], -1))
```
Comment (prady0t) [inline scipy/interpolate/_bary_rational.py:58]: isfinite() already checks for NaNs

File: scipy/interpolate/_bary_rational.py [hunk 7]
```diff
@@ -194,19 +211,24 @@ def roots(self):
...
                                                       self._support_points))
             E[0, 1:] = self.weights * self._support_values
             E[1:, 0] = 1
-            np.fill_diagonal(E[1:, 1:], self._support_points)
-
-            zer = scipy.linalg.eigvals(E, B)
-            self._roots = zer[np.isfinite(zer)]
+            i = xp.arange(E[1:, 1:].shape[0])
+            e = xp.where(i[:, None] == i[None, :], self._support_points, E[1:, 1:])
+            E = xp.concat([E[:1, :], xp.concat([E[1:, :1], e], axis=1)], axis=0)
```
Comment (prady0t) [inline scipy/interpolate/_bary_rational.py]: xp.fill_diagonal() does not exist. I tied to reconstruct the whole array. I would love a review here.

File: scipy/interpolate/_bary_rational.py [hunk 8]
```diff
@@ -731,22 +765,25 @@ class FloaterHormannInterpolator(_BarycentricRational):
...
         n = w.size
         for k in range(n):
             for i in range(max(k-d, 0), min(k+1, n-d)):
-                w[k] += 1/np.prod(np.abs(np.delete(z[k] - z[i : i + d + 1], k - i)))
-        w *= (-1.)**(np.arange(n) - d)
+                diff = z[k] - z[i : i + d + 1]
+                diff = xp.concat([diff[:(k - i)], diff[(k - i) + 1:]], axis=0)
+                w[k] += 1 / xp.prod(xp.abs(diff))
+        one = xp.ones((n,), dtype=w.dtype)
+        w *= xp.where((xp.arange(n) - d) % 2 == 0, one, -one)
```
Comment (prady0t) [inline scipy/interpolate/_bary_rational.py:825]: Two reasons for this:
Python float scalars can only be promoted with floating-point arrays.
array_api_strict() does not allow two scalars in xp.where

File: scipy/interpolate/tests/test_bary_rational.py [hunk 9]
```diff
@@ -315,37 +323,38 @@ def test_iv(self):
...
-        xx = np.linspace(-1, 1, num=1000)
+        r = FloaterHormannInterpolator(z, xp.sin(z), d=12)
+        xx = xp.linspace(-1, 1, num=1000)
         zz = xx + xx*1j
-        assert_allclose(r(zz), np.sin(zz), rtol=1e-12)
+        xp_assert_close(r(zz), xp.sin(zz), rtol=1e-12)
 
-    def test_polyinterp(self):
+    # currently fails for xp test since BarycentricInterpolator is not complaint.
+    def test_polyinterp(self, xp):
```
Comment (prady0t) [inline scipy/interpolate/tests/test_bary_rational.py:368]: Will work on this after this PR, also xp needs to be removed from here.

**Comment by prady0t** [comment]
Hi all,
I have carefully made all the changes and added a few comments. I would love your help and reviews to get this merged. 
Thanks.

File: scipy/interpolate/_bary_rational.py (hunk 7 above)
Comment (prady0t) [reply scipy/interpolate/_bary_rational.py]: I also found this implementation in `array_api_extra`
https://github.com/data-apis/array-api-extra/pull/500

Comment (prady0t) [reply scipy/interpolate/_bary_rational.py]: Also, this in sklearn: https://github.com/scikit-learn/scikit-learn/blob/900adff57900af5350fc545c4d39aceff5361a31/sklearn/utils/_array_api.py#L709

Comment (ev-br) [reply scipy/interpolate/_bary_rational.py]: Please factor this into a private helper (_fill_diagonal or xp_fill_diagonal or some such), and add a comment with a link to the array-api-extra discussion.

Comment (prady0t) [reply scipy/interpolate/_bary_rational.py]: Thanks for the suggestion! I've added it, and it looks much cleaner now; there's no need to concatenate.

**Comment by ev-br** [comment]
This starts to look abandoned, any chance you're planning to return to this PR @prady0t ?

**Comment by prady0t** [comment]
Yes. Have been traveling, will be working on this soon

**Comment by j-bowhay** [comment]
@prady0t I would like to apologies for my earlier comments they were far too strongly worded. I'll try to take another look at this soon

**Comment by prady0t** [comment]
> I would like to apologies for my earlier comments they were far too strongly worded.

No worries! I did a bad job describing the PR anyway😅 

> I'll try to take another look at this soon

I've pushed some commits that should fix the failing test cases. I would love your reviews.

File: scipy/interpolate/_bary_rational.py [hunk 10]
```diff
@@ -170,12 +196,13 @@ def residues(self):
                                       " data.")
         if self._residues is None:
             # Compute residues via formula for res of quotient of analytic functions
-            with np.errstate(divide="ignore", invalid="ignore"):
```
Comment (prady0t) [inline scipy/interpolate/_bary_rational.py:173]: @j-bowhay 
Currently, the cause of the CI failure is replacing this with the ignore filter. Would it be okay to add this file to `test_warning_calls_filters`?
